bap-vibes/src/patch_c.ml

```diff
@@ -0,0 +1,1501 @@
+  | Cabs.OR -> translate_short_circuit_or lhs rhs
+  | Cabs.BAND -> default LAND ~no_ptr:true
+  | Cabs.BOR -> default LOR ~no_ptr:true
+  | Cabs.XOR -> default XOR ~no_ptr:true
+  | Cabs.SHL -> default SHL ~no_ptr:true
+  | Cabs.SHR -> default SHR ~no_ptr:true
+  | Cabs.EQ -> default EQ
+  | Cabs.NE -> default NE
+  | Cabs.LT -> default LT
+  | Cabs.GT -> default GT
+  | Cabs.LE -> default LE
+  | Cabs.GE -> default GE
+  | Cabs.ASSIGN -> translate_assign lhs rhs
+  | Cabs.ADD_ASSIGN -> translate_compound ADD b lhs rhs
+  | Cabs.SUB_ASSIGN -> translate_compound SUB b lhs rhs
+  | Cabs.MUL_ASSIGN -> translate_compound MUL b lhs rhs ~no_ptr:true
+  | Cabs.DIV_ASSIGN -> translate_compound DIV b lhs rhs ~no_ptr:true
+  | Cabs.MOD_ASSIGN -> translate_compound MOD b lhs rhs ~no_ptr:true
+  | Cabs.BAND_ASSIGN -> translate_compound LAND b lhs rhs ~no_ptr:true
+  | Cabs.BOR_ASSIGN -> translate_compound LOR b lhs rhs ~no_ptr:true
+  | Cabs.XOR_ASSIGN -> translate_compound XOR b lhs rhs ~no_ptr:true
+  | Cabs.SHL_ASSIGN -> translate_compound SHL b lhs rhs ~no_ptr:true
+  | Cabs.SHR_ASSIGN -> translate_compound SHR b lhs rhs ~no_ptr:true
+
+(* Captures the short-circuiting semantics of the && operator. *)
+and translate_short_circuit_and
+    (lhs : Cabs.expression)
+    (rhs : Cabs.expression) : (stmt * exp option * stmt) transl =
+  let exp = translate_expression_strict "translate_short_circuit_and" in
+  let* bits =
+    Transl.gets @@ fun {target; _} -> Theory.Target.bits target in
+  let* spre1, e1, spost1 = exp lhs in
+  let t1 = typeof e1 in
+  let* spre2, e2, spost2 = exp rhs in
+  let t2 = typeof e2 in
+  match typ_unify_ptr_to_int bits t1 t2 with
+  | None -> typ_unify_error Cabs.(BINARY (AND, lhs, rhs)) t1 t2
+  | Some t ->
+    let e1 = with_type e1 t in
+    let e2 = with_type e2 t in
+    let+ tmp = new_tmp t in
+    let eff = sequence [
+        spre1;
+        ASSIGN (tmp, e1);
+        spost1;
+        IF (
+          VARIABLE tmp,
+          sequence [
+            spre2;
+            ASSIGN (tmp, e2);
+            spost2
+          ],
+          NOP);
+      ] in
+    eff, Some (VARIABLE tmp), NOP
+
+(* Captures the short-circuiting semantics of the || operator. *)
+and translate_short_circuit_or
+    (lhs : Cabs.expression)
+    (rhs : Cabs.expression) : (stmt * exp option * stmt) transl =
+  let exp = translate_expression_strict "translate_short_circuit_or" in
+  let* bits =
+    Transl.gets @@ fun {target; _} -> Theory.Target.bits target in
+  let* spre1, e1, spost1 = exp lhs in
+  let t1 = typeof e1 in
+  let* spre2, e2, spost2 = exp rhs in
+  let t2 = typeof e2 in
+  match typ_unify_ptr_to_int bits t1 t2 with
+  | None -> typ_unify_error Cabs.(BINARY (OR, lhs, rhs)) t1 t2
+  | Some t ->
+    let e1 = with_type e1 t in
+    let e2 = with_type e2 t in
+    let+ tmp = new_tmp t in
+    let eff = sequence [
+        spre1;
+        ASSIGN (tmp, e1);
+        spost1;
+        IF (
+          VARIABLE tmp,
```

Comment:
maybe I am missing something, but the implementation is the same as with `&&`, shouldn't the variable be negated here?

Reply:
The arms of the `IF` are swapped in each case. If the first condition of the `||` is true then we can just `NOP` and return the value in `tmp`.